core/src/main/java/org/jboss/pnc/rex/facade/TaskProviderImpl.java

```diff
@@ -103,6 +103,14 @@ public void cancel(String taskName) {
         controller.setMode(taskName, Mode.CANCEL);
     }
 
+    @Override
+    @Transactional
+    public void cancelByCorrelationID(String correlationID) {
+        registry.getTasksByCorrelationID(correlationID).stream()
```

Comment:
I wonder if it'd be better to cancel just the leaf nodes from returned list of tasks. It'd be more predictable, logs would be cleaner and the algorithm is already recursive (from leaf to roots) so it is enough.

Reply:
Makes sense! Is there an easy way to iterate through the tasks in order of running execution?

Reply:
I meant to cancel **just** leaves (dependant task will get recursed) but there's no ultra straightforward way to do this because a dependency of a Task is not required to be in the same correlation and the dependency wouldn't be returned by `registry.getTasksByCorrelationID(correlationID)` and the Task itself wouldn't be a leaf and therefore not cancelled with this method.

Technically, semi-easy way could be to use google commons graph, add just those edges where both source+target is in the list from `registry.getTasksByCorrelationID(correlationID)` and then search by number of incident/outgoing edges equal to 0 to get leaves. That could work.